src/forms/upload/async-file-saver.service.js

```
@@ -0,0 +1,46 @@
+
+class AsyncFileSaver {
+  constructor($http) {
+    this.$http = $http;
+  }
+
+  save(fieldName, file, httpOptions) {
+    if (!httpOptions) {
+      throw new Error('You must supply httpOptions');
+    }
+    const formData = new FormData();
+    formData.append(fieldName, file);
+
+    const $httpOptions = prepareHttpOptions(httpOptions);
+
+    const method = $httpOptions.method ? $httpOptions.method : 'POST';
+    delete $httpOptions.method;
+
+    // For testing
+    if (method === 'GET') {
+      return this.$http.get($httpOptions.url, $httpOptions);
+    }
```

Can we please remove this method related logic from the final service? We should never try to upload data using GET - if we want this to be really strict, we should throw an error here when the passed `$httpOptions` have a `GET` method property.

---

The reason I had this was so that you try this from the demo page, where we just fetch a JSON file to represent the response.  Not sure of another easy way to do this short of a custom web server to provide that response?

---

You can override the service in the demo application by using the same injection token (name of the service) with a different implementation.

```js
class AsyncFileSaverDemoMock {
  save(fieldName, file, httpOptions) {
    // Do whatever needed here for demo to work
  }

angular.module('my.demo.module').service('AsyncFileSaver',  AsyncFileSaverDemoMock); 
```

If this code is executed after the original declaration of the `AsyncFileSaver`, you simple override the value referenced by the injection token and the injector will return your mocked instance in your demo application. 

This way the "production" service would contain only the necessary code for sending the request while the mocked service for demo purposes can implement the same interface and do whatever needed for the demo purposes.

I don't think this is hacky, as AngularJs works this way and will enable you to remove unnecessary code only for testing purposes from production code.

---

Good idea, I wrote this before I extracted the AsyncFileSaver, when it would have been more challenging.